File: week3/8bitbully/reverseList.cpp
// 206. Reverse Linked List

struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
  using N = ListNode;

public:
  ListNode *reverseList(ListNode *head) {
    N *reverse = nullptr;
    N *current = head;

    while (current) {
      N *dummy = current->next;
      current->next = reverse;
      reverse = current;
      current = dummy;
    }

    return reverse;
  }
};

File: week3/MarshalLee/ReverseList.cs
using System;

public static ListNode ReverseList(ListNode head)
{
    ListNode previousNode = null;
    while (head != null)
    {
        ListNode currNext = head.next;
        head.next = previousNode;
        previousNode = head;
        head = currNext;
    }
    return previousNode;
    //ListNode previous = null;
    //ListNode curr = head;
    //while(curr != null)
    //{
    //    ListNode next = curr.next;
    //    curr.next = previous;
    //    previous = curr;
    //    curr = next;
    //}
    //return previous;
}